File: EmailCalendarsClient/MailSender/AadGraphApiDelegatedClient.cs
using Microsoft.Graph;
using Microsoft.Identity.Client;
using System.Collections.Generic;
using System.Configuration;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using GraphEmailClient;

namespace EmailCalendarsClient.MailSender
{
    public class AadGraphApiDelegatedClient
    {
        private readonly HttpClient _httpClient = new HttpClient();
        private IPublicClientApplication _app;

        private static readonly string AadInstance = ConfigurationManager.AppSettings["AADInstance"];
        private static readonly string Tenant = ConfigurationManager.AppSettings["Tenant"];
        private static readonly string ClientId = ConfigurationManager.AppSettings["ClientId"];
        private static readonly string Scope = ConfigurationManager.AppSettings["Scope"];

        private static readonly string Authority = string.Format(CultureInfo.InvariantCulture, AadInstance, Tenant);
        private static readonly string[] Scopes = { Scope };

        public void InitClient()
        {
            _app = PublicClientApplicationBuilder.Create(ClientId)
                .WithAuthority(Authority)
                .WithRedirectUri("http://localhost:65419") // needed only for the system browser
                .Build();

            TokenCacheHelper.EnableSerialization(_app.UserTokenCache);
        }

        public async Task<IAccount> SignIn()
        {
            var accounts = (await _app.GetAccountsAsync()).ToList();
            try
            {
                var result = await _app.AcquireTokenSilent(Scopes, accounts.FirstOrDefault())
                    .ExecuteAsync()
                    .ConfigureAwait(false);

                return result.Account;
            }
            catch (MsalUiRequiredException)
            {
                var builder = _app.AcquireTokenInteractive(Scopes)
                    .WithAccount(accounts.FirstOrDefault())
                    .WithUseEmbeddedWebView(false)
                    .WithPrompt(Microsoft.Identity.Client.Prompt.SelectAccount);

                var result = await builder.ExecuteAsync().ConfigureAwait(false);

                return result.Account;
            }
        }

        public async Task<AuthenticationResult> AcquireTokenSilent()
        {
            var accounts = await GetAccountsAsync();
            var result = await _app.AcquireTokenSilent(Scopes, accounts.FirstOrDefault())
                    .ExecuteAsync()
                    .ConfigureAwait(false);

            return result;
        }          

        public async Task<IList<IAccount>> GetAccountsAsync()
        {
            var accounts = await _app.GetAccountsAsync();
            return accounts.ToList();
        }

        public async Task RemoveAccountsAsync()
        {
            IList<IAccount> accounts = await GetAccountsAsync();

            // Clears the library cache. Does not affect the browser cookies.
            while (accounts.Any())
            {
                await _app.RemoveAsync(accounts.First());
                accounts = await GetAccountsAsync();
            }
        }

        public async Task SendEmailAsync(Message message)
        {
            var result = await AcquireTokenSilent();

            _httpClient.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", result.AccessToken);
            _httpClient.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));

            GraphServiceClient graphClient = new GraphServiceClient(_httpClient)
            {
                AuthenticationProvider = new DelegateAuthenticationProvider(async (requestMessage) =>
                {
                    requestMessage.Headers.Authorization = new AuthenticationHeaderValue("Bearer", result.AccessToken);
                })
            };

            var saveToSentItems = true;

            await graphClient.Me
                .SendMail(message, saveToSentItems)
                .Request()
                .PostAsync();
        }

    }
}

File: VbEmailClient/MailSender/AadGraphApiDelegatedClient.vb
Imports Microsoft.Graph
Imports Microsoft.Identity.Client
Imports System.Configuration
Imports System.Globalization
Imports System.Net.Http
Imports System.Net.Http.Headers
Imports VbEmailClient.GraphEmailClient

Namespace EmailCalendarsClient.MailSender
    Public Class AadGraphApiDelegatedClient
        Private ReadOnly _httpClient As HttpClient = New HttpClient()
        Private _app As IPublicClientApplication
        Private Shared ReadOnly AadInstance As String = ConfigurationManager.AppSettings("AADInstance")
        Private Shared ReadOnly Tenant As String = ConfigurationManager.AppSettings("Tenant")
        Private Shared ReadOnly ClientId As String = ConfigurationManager.AppSettings("ClientId")
        Private Shared ReadOnly Scope As String = ConfigurationManager.AppSettings("Scope")
        Private Shared ReadOnly Authority As String = String.Format(CultureInfo.InvariantCulture, AadInstance, Tenant)
        Private Shared ReadOnly Scopes As String() = {Scope}

        Public Sub InitClient()
            _app = PublicClientApplicationBuilder.Create(ClientId).WithAuthority(Authority).WithRedirectUri("http://localhost:65419").Build()
            TokenCacheHelper.EnableSerialization(_app.UserTokenCache)
        End Sub

        Public Async Function SignIn() As Task(Of IAccount)
            Dim accounts = (Await _app.GetAccountsAsync()).ToList()

            Try
                Dim result = Await _app.AcquireTokenSilent(Scopes, accounts.FirstOrDefault()).ExecuteAsync().ConfigureAwait(False)
                Return result.Account
            Catch __unusedMsalUiRequiredException1__ As MsalUiRequiredException
                Dim builder = _app.AcquireTokenInteractive(Scopes).WithAccount(accounts.FirstOrDefault()).WithUseEmbeddedWebView(False).WithPrompt(Microsoft.Identity.Client.Prompt.SelectAccount)
                Dim result = builder.ExecuteAsync().GetAwaiter().GetResult()
                Return result.Account
            End Try
        End Function

        Public Async Function AcquireTokenSilent() As Task(Of AuthenticationResult)
            Dim accounts = Await GetAccountsAsync()
            Dim result = Await _app.AcquireTokenSilent(Scopes, accounts.FirstOrDefault()).ExecuteAsync().ConfigureAwait(False)
            Return result
        End Function

        Public Async Function GetAccountsAsync() As Task(Of IList(Of IAccount))
            Dim accounts = Await _app.GetAccountsAsync()
            Return accounts.ToList()
        End Function

        Public Async Function RemoveAccountsAsync() As Task
            Dim accounts As IList(Of IAccount) = Await GetAccountsAsync()

            While accounts.Any()
                Await _app.RemoveAsync(accounts.First())
                accounts = Await GetAccountsAsync()
            End While
        End Function

        Public Async Function SendEmailAsync(ByVal message As Message) As Task
            Dim result = Await AcquireTokenSilent()
            _httpClient.DefaultRequestHeaders.Authorization = New AuthenticationHeaderValue("Bearer", result.AccessToken)
            _httpClient.DefaultRequestHeaders.Accept.Add(New MediaTypeWithQualityHeaderValue("application/json"))
            Dim graphClient As GraphServiceClient = New GraphServiceClient(_httpClient) With {
                .AuthenticationProvider = New DelegateAuthenticationProvider(Async Function(requestMessage)
                                                                                 requestMessage.Headers.Authorization = New AuthenticationHeaderValue("Bearer", result.AccessToken)
                                                                             End Function)
            }
            Dim saveToSentItems = True
            Await graphClient.[Me].SendMail(message, saveToSentItems).Request().PostAsync()
        End Function
    End Class
End Namespace
